### ISSUE TITLE
DRY up the Ash code interface 

### ISSUE THREAD

**Issue by ahacking** [ash-project/ash#773]
**Describe the problem**

When specifying a code interface for our actions we are required to specify the argument list which seems redundant when the argument list is already specified in detail on the action definition.

**Describe the solution you'd like**

I would like Ash to infer the argument list from the action as it only introduces the possibility of an inconsistency with our actions.

**Express the feature either with a change to resource syntax, or with a change to the resource interface**

For example, just specify the *list* of actions you want it expose and "infer the rest."

```elixir
code_interface do
  define_for Helpdesk.Support

  define_actions [:open, :close, :read]
end
```

**Describe alternatives you've considered**

I probably prefer removing the code interface block completely and move code interface directives into the action (or calculation) block:

```elixir
actions do
  # this includes all actions in the block by default
  code_interface Account.User

  # can also "publish" in other code interfaces and also explicity list the exposed actions
  code_interface Account.User.Mutating, [:create, :update, :destroy, :register]

  defaults [:create, :read, :update, :destroy]

  create :register do
    accept [:email]
    # this registers this action in a different module under a different function as we may also want to collect registration types from other resources :register_admin, :register_tenant and so on..
    code_interface Account.Registration, as: :register_user

    argument :password, :string do
      allow_nil? false
    end

    argument :password_confirmation, :string do
      allow_nil? false
    end
  end
end
```

**Comment by zachdaniel** [comment]
We've discussed this/things like this. There are some issues with using the list of arguments, specifically that it makes the argument order matter both in the block *and* the order in the `accept` list of actions, and would make it hard to compose the two. For instance:

```elixir
create :register do
  accept [:email]

  argument :password, :string

  ...
end
```

if you wanted `Resource.create(email, password)`, we'd have to put `accept` before `argument` when building the interface. But if you wanted `Resource.create(password, email)`, we'd have to put it after. There isn't really any way to figure out code interface argument order otherwise. With that in mind it seems like it would be more ergonomic *at first* and then far less ergonomic when trying to maintain a specific function signature while making changes to your action. I'm going to close this for now, but feel free to continue the discussion, happy to have my mind changed :)

**Comment by ahacking** [comment]
Thanks for considering.

I was assuming the argument order in the action was sufficient, but you do have valid points.

From a maintenance perspective removing an argument would be a breaking change regardless, as would be adding a non optional mandatory argument. Adding optional args to the end would not be breaking for the current approach and the proposed, but inserting an argument between existing ones or changing the order of arguments in the action block would be breaking for the new approach.

A possible "way out" where you absolutely must preserve the API is simply define another action which would have the breaking argument definition with different arguments, preserving the original.

I think we can do better however.

Another "way out" is to support overriding the argument order as an option if the action code_interface directive was adopted as the way forward. Since you can call code_interface multiple times in my proposal you can define an API different ways and in multiple modules if need be and preserve backward compatability.

I would probably also consider the ultimate escape hatch here for code_interface and just let the user write whatever function head and body calling convention they desire within a code_interface do end block, perhaps they do want a query as the first argument to their special read action ...

Thoughts?

**Comment by zachdaniel** [comment]
If you have to specify the action order in any non-trivial case then ultimately the code looks the same except for being able to do it in-line in the action right?

i.e
```elixir
create :create do
  code_interface ....
end
```

At which point its only a marginal improvement over what we have now right?

**Comment by zachdaniel** [comment]
Not even necessarily sure it's an improvement, as now you can't easily look at "what is in the code interface". You have to scan each action. In the same way we don't define graphql mutations or queries directly inside of an action.

**Comment by ahacking** [comment]
> If you have to specify the action order in any non-trivial case then ultimately the code looks the same except for being able to do it in-line in the action right?
> 
> i.e
> 
> ```elixir
> create :create do
>   code_interface :name, args: [:foo, :bar, :baz]
> end
> ```
> 
> At which point its only a marginal improvement over what we have now right?

That's if you actually  _have_ to change the argument order, but I would argue why not simply define things in the desired order so the need to override is more of an edge case.

**Comment by ahacking** [comment]
> Not even necessarily sure it's an improvement, as now you can't easily look at "what is in the code interface". You have to scan each action. In the same way we don't define graphql mutations or queries directly inside of an action.

That is true, but you do have to look carefully at your actions in any event when defining a code interface to be sure you included all the attributes and arguments you need to call the action.

It just _feels_ a bit like the old days of C programming having prototypes in a separate place like in headers and making sure you match the .h function prototype with the .c function definition.

**Comment by zachdaniel** [comment]
> That's if you actually have to change the argument order, but I would argue why not simply define things in the desired order so the need to override is more of an edge case.

The problem is that `accept` is a single list. So if you have something like this:

```elixir
accept [:foo, :bar, :baz]
argument :buz, ...
```

and you want `function(foo, buz, bar, baz)`

**Comment by zachdaniel** [comment]
I'm open to adding something like this to actions:

```elixir
create :foo do
  define :function, args: [:foo, :bar]
end
```

Just not automatically deriving the args from the action definition.

**Comment by zachdaniel** [comment]
If you want to colocate the definition, that is

**Comment by ahacking** [comment]
> I'm open to adding something like this to actions:
> 
> ```elixir
> create :foo do
>   define :function, args: [:foo, :bar]
> end
> ```
> 
> Just not automatically deriving the args from the action definition.

I honestly don't think there is sufficient value in having two places to express code interfaces if they both require explicit arg lists.

The split between accept directives (being a list) and arguments being singular directives kind of ruins the idea of an ordering.

I think we can retire this idea, but thanks for giving it  consideration.
